File: Binary_search/lib/binary_search_tree.rb
# There are many ways to implement these methods, feel free to add arguments 
# to methods as you see fit, or to create helper methods.
require_relative "bst_node"
class BinarySearchTree
  attr_accessor :root
  def initialize
    @root = nil
  end

  def insert(value)
    if @root.nil?
      @root = BSTNode.new(value) 
      return
    end 
    handle_insert(@root,value)
  end



  def find(value, tree_node = @root)
    return nil if tree_node.nil?
    return tree_node if tree_node.value == value
    
    if tree_node.value > value 
      find(value, tree_node.left)
    else 
      find(value, tree_node.right)
    end 
  end

  def delete(value)

    node = find(value)
    parent = find_parent(value)
    if node.left.nil? && node.right.nil? #has no children
      @root = nil if node == root
      parent.left = nil if parent.left == node 
      parent.right = nil if parent.right == node
    elsif node.left && node.right.nil? #has left children
      parent.left = parent.left.left if parent.left == node 
      parent.right = parent.right.left if parent.right == node
    elsif node.right && node.left.nil? #has right children
      parent.left = parent.left.right if parent.left == node 
      parent.right = parent.right.right if parent.right == node
    else 
      remove_node(node, parent)
    end

  end

  def remove_node(node, node_parent)
    max = maximum(node.left)
    parent = find_parent(max.value)
    if max.left ##if max node has left child, promote it
      parent.right = max.left 
    end 
    node_parent.left = max if node_parent.left == node  #reassign node to be max node
    node_parent.right = max if node_parent.right == node
  end
  # helper method for #delete:
  def maximum(tree_node = @root)
    return tree_node if tree_node.right.nil?
    maximum(tree_node.right)
  end

  def depth(tree_node = @root)
    if tree_node.nil?
      return -1
    end 
    left_depth = depth(tree_node.left)
    left_depth += 1
    right_depth = depth(tree_node.right)
    right_depth += 1
    return left_depth > right_depth ? left_depth : right_depth 
  end 

  def is_balanced?(tree_node = @root)
    return true if tree_node.nil?
    left_depth = depth(tree_node.left)
    right_depth = depth(tree_node.right)

    left_balanced = is_balanced?(tree_node.left)
    right_balanced = is_balanced?(tree_node.right)
    return true if (left_depth-right_depth).abs <= 1 && left_balanced && right_balanced
    false
  end

  def in_order_traversal(tree_node = @root, arr = [])
    return if tree_node.nil?
    in_order_traversal(tree_node.left, arr)
    arr.push(tree_node.value)
    in_order_traversal(tree_node.right, arr)
    arr
  end


  private
  # optional helper methods go here:
    def has_children(node)
      node.right || node.left
    end 

    def handle_insert(node,value)
      if node.nil?
        node = BSTNode.new(value)
        return
      end 

      if node.value > value
        if node.left.nil?
          node.left = BSTNode.new(value)
          return
        end
        handle_insert(node.left, value) unless node.left.nil?
      end 

      if node.value < value 
        if node.right.nil?
          node.right = BSTNode.new(value)
          return 
        end 
        handle_insert(node.right, value) 
      end 
    end 

    def find_parent(value, tree_node = @root)
      return nil if tree_node.nil?
      return @root if @root.value == value
      return tree_node if tree_node.left.value == value || tree_node.right.value == value
      if tree_node.value > value 
        find_parent(value, tree_node.left)
      else 
        find_parent(value, tree_node.right)
      end 
    end 

end

File: CTCI/Trees/binary_search_tree.js
const BSTNode = require('./bst_node');
class BinarySearchTree {
    constructor() {
        this.root = null;
    }

    insert(val) {
        if (!this.root) {
            this.root = new BSTNode(val);
            return;
        }
        this.handle_insert(this.root,val);
    }
    handle_insert(node=this.root, val) {
        if(!node) {
            node = new BSTNode(val);
            return;
        }
        
        if(node.value > val) {
            if(!node.left) {
                node.left = new BSTNode(val);
                return;
            }
            this.handle_insert(node.left,val);
        }
        if(node.value < val) {
            if(!node.right) {
                node.right = new BSTNode(val);
                return;
            }
            this.handle_insert(node.right,val);
        }
    }
    find(val, node=this.root) {
        if(!node) return null;
        if (node.value === val) {
            return node;
        }
        if(node.value > val) {
            return this.find(val,node.left);
        } else {
            return this.find(val,node.right);
        }
        
    }
    bfs(node=this.root) {
        const result = [];
        let queue = [];
        queue.push(node);
        result.push(node.value);
        while(queue.length !== 0) {
            let el = queue.shift();
            if (el.right) {
                result.push(el.right.value);
                queue.push(el.right);
            }
            if(el.left) {
                result.push(el.left.value);
                queue.push(el.left);
            }
            
        }
        return result;
    }
    in_order_traversal(node = this.root, arr = []) {
        if(!node) return;
        this.in_order_traversal(node.left, arr);
        arr.push(node.value);
        this.in_order_traversal(node.right, arr);
        return arr;
    }
    depth(node=this.root) {
        if(!node) return 0;
        return (Math.max(this.depth(node.left),this.depth(node.right))+1);
    }
    is_balanced(node=this.root) {
        if(!node) return true;
        let left_depth = this.depth(node.left);
        let right_depth = this.depth(node.right);
        let left_balanced = this.is_balanced(node.left);
        let right_balanced = this.is_balanced(node.right);
        if(Math.abs(left_depth - right_depth) <= 1 && left_balanced && right_balanced) {
            return true;
        }
        return false;
    }
}

const search = new BinarySearchTree();
search.insert(3);
search.insert(5);
search.insert(4);
search.insert(2);
search.insert(1);

// console.log(search.in_order_traversal());
// console.log(search.is_balanced());
console.log(search.bfs());